### PR TITLE
Templated functions for **_eq and **_compare

### LINKED ISSUE #529: Templated functions for `*_eq` and `*_compare`
There many functions of the form `*_eq()` and `*_compare()` in the [dict.cpp](https://github.com/sympy/symengine/blob/master/symengine/dict.cpp) file.
Use templated functions instead as we already do for printing, so that the logic abstracts out.

From @certik's comment [here](https://github.com/sympy/symengine/pull/511#discussion_r34426918)

### PR DESCRIPTION
**Pull request by shivamvats**
@certik @Sumith1896 
Fixes #529

### CODE REVIEW

**Comment by Sumith1896** [comment]
Thanks @shivamvats , it's a lot cleaner now.
+1 from my side.

File: symengine/functions.cpp [hunk 1]
```diff
@@ -1703,7 +1703,7 @@ bool Subs::__eq__(const Basic &o) const
 {
     if (is_a<Subs>(o) &&
             eq(*arg_, *(static_cast<const Subs &>(o).arg_)) &&
-            map_basic_basic_eq(dict_, static_cast<const Subs &>(o).dict_))
+            map_eq<map_basic_basic>(dict_, static_cast<const Subs &>(o).dict_))
```
Comment (certik) [inline symengine/functions.cpp]: Does it work if you call it as `map_eq(dict_, static_cast<const Subs &>(o).dict_))`? Many times you don't need to specify the template.

**Comment by certik** [comment]
I think it looks good. The only question is whether we want to hide the templated functions in dict.cpp (and only expose the original non-templated versions that just call the templated ones underneath) to speed up the compilation.

File: symengine/functions.cpp (hunk 1 above)
Comment (shivamvats) [inline symengine/functions.cpp]: Indeed it does! Interesting.
I will make the changes.

**Comment by shivamvats** [comment]
The only argument against doing so is that we'll need to define a function every time we want to use a new type with the templated function. What do you suggest?

**Comment by certik** [comment]
Make the change to remove the template from the call and I guess we can keep it in the header file, it shouldn't be a big deal.

**Comment by Sumith1896** [comment]
What if you pass two different types? Does it compile?
iirc it shouldn't but if it does, we need checks

**Comment by shivamvats** [comment]
@certik Done
@Sumith1896 Why do you think so?

**Comment by Sumith1896** [comment]
I felt it shouldn't compile. If it did, checking dicts are of same type or not would be needed.
Everything's fine now.

**Comment by certik** [comment]
Try it and report back. I think it won't compile.

**Comment by certik** [comment]
This looks good, merging.

**Comment by shivamvats** [comment]
Thanks

**Pull request merged**